### DISCUSSION TITLE
fontconfig on windows troubles with solution

### DISCUSSION THREAD

**Discussion by chrismaster** [FriedrichFroebel/brother_ql_web#20] [Ideas]
While trying to get some kind of dev environment on my windows pc for #19 , I struggled with fontconfig.
fontconfig seems to be an os dependency and you have to install it, if it is missing on the os.
On the net many have troubles with it on windows.
For some with the same problem, this might be some help....

After adding fontra https://pypi.org/project/fontra/
all I had to do was change in 
`font_helpers.py` function `get_fonts` with imports `import fontra` and `from pathlib import Path` to
```
def get_fonts(folder: str | None = None) -> dict[str, dict[str, str]]:
    if folder:
        fontra.FONTDIRS_CUSTOM.append(Path(str))
        fontra.update_custom_fontfiles_index()
        fontra.update_fontrefs_index()
    else:
        fontra.init_fontdb()
    fonts: dict[str, dict[str, str]] = defaultdict(dict)
    families = fontra.all_fonts()
    for family in families:
        styles = fontra.get_font_styles(family)
        for style in styles:
            fonts[family][style] = fontra.get_font(family, style).path.absolute().as_posix()
    return dict(fonts)
```

It should do its job on macos and linux, but didn't test it.

**Comment by FriedrichFroebel** [comment]
I have never heard of `fontra` and it seems to be a rather new and unknown package. The first thing I stumbled upon is that it requires at least Python 3.9 - while Python 3.8 is EOL soon, this would be a breaking change.

For Windows, Anaconda might be an option, which has pre-built binaries for Windows as well: https://anaconda.org/anaconda/fontconfig (You might be able to use the binaries for a manual installation as well, but according to the `info/index.json` file, it seems like you will need some further dependencies which involves further work.)

**Comment by chrismaster** [comment]
I tried fontconfig from anaconda, from mingw, and some more, but didn't manage to make it work. So I gave fontra a try. Since it worked, I thought it might help someone with troubles on windows too. On linux and in container, everything is fine with fontconfig. Since I have a container running brother_ql_web, I am fine with dep on fontconfig.

ATM I am learning how this python universe works, my programming skills are 0, so fontra was a nice success. Setuptools, poetry, bottle, jinga so many things I never used, never heard of and hope someday I have a PR with tests for #19 . :-)
Thx for the help.

**Reply by FriedrichFroebel** [reply]
In the meantime, I am evaluating whether I can provide *fontra* as a drop-in replacement for *fontconfig*. For now, the results generated by *fontra* miss some entries to pass the existing integration tests, but maybe upstream has an idea about this: https://github.com/NCBM/fontra/issues/3

**Comment by FriedrichFroebel** [comment]
The *master* branch now provides an optional fallback to `fontra`.
